File: backend/services/token_mode.py
"""
Token mode helper — single source of truth for "is this a simulated (SIM_)
token or a real on-chain token?"

All backend responses that return token_mint_address should carry
`is_simulated` and `token_mode` fields derived from this helper so the
frontend never has to string-match `SIM_` directly.

Do NOT use this helper for DUM Points. DUM Points are a separate system
(real Stripe on-ramp + optional on-chain claim) and are out of scope for
the Door A (simulated-token labeling) PR.
"""

from __future__ import annotations

from typing import Literal


TokenMode = Literal["simulated", "on_chain"]


def is_simulated_token(token_mint_address: str | None) -> bool:
    """
    True when the mint address is missing or is a simulated placeholder.

    The ``SIM_`` prefix is written by backend/api/routes/token.py and
    backend/api/routes/launch.py at project-creation time. A missing
    mint is also treated as simulated — there is no live token to point
    at yet, so callers should never advertise live-market behavior.
    """
    if not token_mint_address:
        return True
    return token_mint_address.startswith("SIM_")


def token_mode(token_mint_address: str | None) -> TokenMode:
    """Return 'simulated' or 'on_chain' for a given mint address."""
    return "simulated" if is_simulated_token(token_mint_address) else "on_chain"

File: frontend/lib/tokenMode.ts
/**
 * Token mode helper — single source of truth for "is this a simulated (SIM_)
 * token or a real on-chain token?"
 *
 * Every frontend surface that renders a per-project token MUST route through
 * this helper. Do not string-match "SIM_" in components. If you add a new
 * place that displays a mint address, price, market cap, trade button, or
 * candle chart, wire it through isSimulatedToken(...) so a future switch to
 * real on-chain tokens is a one-line change.
 *
 * DUM Points are a separate system (real Stripe on-ramp, optional on-chain
 * claim) and are out of scope. Do not call this helper for DUM Points.
 */

export type TokenMode = "simulated" | "on_chain";

/**
 * True when the mint address is missing, empty, or is a simulated placeholder.
 * The "SIM_" prefix is written by backend/api/routes/token.py and launch.py.
 */
export function isSimulatedToken(mint?: string | null): boolean {
  if (!mint) return true;
  return mint.startsWith("SIM_");
}

export function tokenMode(mint?: string | null): TokenMode {
  return isSimulatedToken(mint) ? "simulated" : "on_chain";
}

/** Short visible label for badges. "DEMO" for simulated, "LIVE" for real. */
export function tokenModeLabel(mint?: string | null): "DEMO" | "LIVE" {
  return isSimulatedToken(mint) ? "DEMO" : "LIVE";
}
